tools/python/boutiques/localExec.py

```diff
@@ -411,56 +410,66 @@ def prepare(self):
+                imageDir = "./"
 
             # Check if container already exists
-            if self._singConExists(conName):
-                return (conName, "Local ({0})".format(conName))
+            if self._singConExists(conName, imageDir):
+                conPath = op.abspath(op.join(imageDir, conName))
+                return conPath, "Local ({0})".format(conName)
 
             # Container does not exist, try to pull it
-            lockdir = conName + "-lock"
+            if self.imagePath:
+                lockDir = self.imagePath + "-lock"
+            else:
+                lockDir = conName + "-lock"
+
             maxcount = 36
             count = 0
-            try:
-                while count < maxcount:
-                    count += 1
-                    try:
-                        os.mkdir(lockdir)
-                    except OSError:
-                        time.sleep(5)
-                    else:
-                        # Check if container was created while waiting
-                        if self._singConExists(conName):
-                            return (conName, "Local ({0})".format(conName))
-                        # Container still does not exist, so pull it
-                        return self._pullSingImage(conName, conIndex, conImage)
-                # If loop times out, check again for existence, otherwise
-                # raise an error
-                if self._singConExists(conName):
-                    return (conName, "Local ({0})".format(conName))
-                raise_error(ExecutorError, "Unable to retrieve Singularity "
-                            "image.")
-            finally:
-                os.rmdir(lockdir)
-                if "SINGULARITY_PULLFOLDER" in os.environ:
-                    del os.environ["SINGULARITY_PULLFOLDER"]
+
+            while count < maxcount:
+                count += 1
+                try:
+                    os.mkdir(lockDir)
+                except OSError:
+                    time.sleep(5)
+                else:
+                    # Check if container was created while waiting
+                    if self._singConExists(conName, imageDir):
+                        self._cleanUpAfterSingPull(lockDir)
+                        conPath = op.abspath(op.join(imageDir, conName))
+                        return conPath, "Local ({0})".format(conName)
+                    # Container still does not exist, so pull it
+                    return self._pullSingImage(conName, conIndex, conImage,
+                                               imageDir, lockDir)
+            # If loop times out, check again for existence, otherwise
+            # raise an error
+            if self._singConExists(conName, imageDir):
+                conPath = op.abspath(op.join(imageDir, conName))
+                return conPath, "Local ({0})".format(conName)
+            raise_error(ExecutorError, "Unable to retrieve Singularity "
+                        "image.")
```

Comment:
Could you make sure that all the lines in this loop are covered by tests?

Reply:
Ok, it's a bit tricky since we can't really have 2 processes running in parallel in a test, but I think I can do it with mocks.